testtime/testtime.go

```diff
@@ -0,0 +1,292 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2024 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+// Package testtime provides a wrapper around time.Timer and its associated
+// functions so that timers can be mocked in tests.
+package testtime
+
+import (
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/snapcore/snapd/testutil"
+)
+
+// Timer is an interface which wraps time.Timer so that it may be mocked.
+//
+// Timer is fully compatible with time.Timer, except that since interfaces
+// cannot have instance variables, we must expose the C channel as a method.
+// Therefore, when replacing a time.Timer with testtime.Timer, any direct
+// accesses of C must be replaced with C().
+//
+// For more information about time.Timer, see: https://pkg.go.dev/time#Timer
+type Timer interface {
```

Comment:
I don't think anyone will import a test package to use the interface, but what you could do is simply make sure that the fake timer and stdlib timer match the interface so that both look similar enough in the context of the test code may want to do.

```
var _ Timer = (*TestTimer)(nil)
var _ Timer = (*time.Timer)(nil)
```

Reply:
That was the original thought behind the package (see commits prior to [e196095](https://github.com/canonical/snapd/pull/14672/commits/e19609531274f5fc448ec5f64e5ea2b32163b721)), but then production code which uses the `C` field of `time.Timer` cannot be replaced by a `TestTimer` nor the `testtime.Timer` interface, as interfaces cannot have instance variables.

And even if `time.Timer` implemented the `testtime.Timer` interface, in order to have production code which uses `time.Timer` use a mockable timer instead, wouldn't they have to use `testtime.Timer` as the type anyway, and just replace `time.AfterFunc` with `testtime.AfterFunc` etc.? There's no way to have a variable of type `*time.Timer` populated with a `*testtime.TestTimer`

Reply:
Either the package locally defines a Timer interface using the one here as a reference (i.e. makes a copy), or the Timer interface could be moved to eg. timeutil which other packages can import. Various `*test` packages are meant to provide things you need for testing only.

Just for the record, we have a single place right now where the Timer interface is used, so I don't mind simply keeping it in the prompting code until we have more opportunities to use it (and the test helpers)